state/global_state.go

```diff
@@ -0,0 +1,341 @@
+func (self *StateDB) setError(err error) {
+	if self.dbErr == nil {
+		self.dbErr = err
+	}
+}
+
+func (self *StateDB) Error() error {
+	return self.dbErr
+}
+
+// Exist reports whether the given account address exists in the state.
+// Notably this also returns true for suicided accounts.
+func (self *StateDB) Exist(addr common.Address) bool {
+	return self.getStateObj(addr) != nil
+}
+
+
+// Empty returns whether the state object is either non-existent
+// or empty according to the EIP161 specification (balance = nonce = code = 0)
+func (self *StateDB) Empty(addr common.Address) bool {
+	so := self.getStateObj(addr)
+	return so == nil || so.empty()
+}
+
+
+// Retrieve the balance from the given address or 0 if object not found
+func (self *StateDB) GetBalance(addr common.Address) *big.Int {
+	StateObj := self.getStateObj(addr)
+	if StateObj != nil {
+		return StateObj.Balance()
+	}
+	return common.Big0
+}
+
+func (self *StateDB) GetNonce(addr common.Address) uint64 {
+	StateObj := self.getStateObj(addr)
+	if StateObj != nil {
+		return StateObj.Nonce()
+	}
+
+	return 0
+}
+
+/*
+ * SETTERS
+ */
+
+// AddBalance adds amount to the account associated with addr.
+func (self *StateDB) AddBalance(addr common.Address, amount *big.Int) {
+	stateObj := self.GetOrNewStateObj(addr)
+	if stateObj != nil {
+		stateObj.AddBalance(amount)
+	}
+}
+
+// SubBalance subtracts amount from the account associated with addr.
+func (self *StateDB) SubBalance(addr common.Address, amount *big.Int) {
+	StateObj := self.GetOrNewStateObj(addr)
+	if StateObj != nil {
+		StateObj.SubBalance(amount)
+	}
+}
+
+func (self *StateDB) SetBalance(addr common.Address, amount *big.Int) {
+	stateObj := self.GetOrNewStateObj(addr)
+	if stateObj != nil {
+		stateObj.SetBalance(amount)
+	}
+}
+
+func (self *StateDB) SetNonce(addr common.Address, nonce uint64) {
+	stateObj := self.GetOrNewStateObj(addr)
+	if stateObj != nil {
+		stateObj.SetNonce(nonce)
+	}
+}
+
+
+//
+// Setting, updating & deleting state object methods.
+//
+
+// updateStateObj writes the given object to the trie.
+func (self *StateDB) updateStateObj(StateObj *StateObj) {
+	addr := StateObj.Address()
+	data, err := rlp.EncodeToBytes(StateObj)
+	if err != nil {
+		panic(fmt.Errorf("can't encode object at %x: %v", addr[:], err))
+	}
+	self.setError(self.globalTrie.TryUpdate(addr[:], data))
+}
+
+// Retrieve a state object given by the address. Returns nil if not found.
+func (self *StateDB) getStateObj(addr common.Address) (StateObj *StateObj) {
+	// Prefer 'live' objects.
+	if obj := self.stateObjects[addr]; obj != nil {
+		/*if obj.deleted {
+			return nil
+		}*/
+		return obj
+	}
+
+	// Load the object from the database.
+	enc, err := self.globalTrie.TryGet(addr[:])
+	if len(enc) == 0 {
+		self.setError(err)
+		return nil
+	}
+	var data Account
+	if err := rlp.DecodeBytes(enc, &data); err != nil {
+		log.Error("Failed to decode state object", "addr", addr, "err", err)
```

Comment:
is it ok that this returns nil because of serialization error but no error was returned or set on StateObj ?